Add example for 3D adaptive histogram equalization (AHE)

m-albert opened the pull request:
## Description

New gallery example illustrating functionality added in this PR https://github.com/scikit-image/scikit-image/pull/4598.

The example compares the results of applying AHE to a 3D image and a degraded version of it to illustrate how AHE can be used to recover contrast and normalize images.


![image](https://user-images.githubusercontent.com/12528388/81348033-9b7b7700-90bd-11ea-9be5-73ffa5be3e90.png)

## Checklist

<!-- It's fine to submit PRs which are a work in progress! -->
<!-- But before they are merged, all PRs should provide: -->
- [ ] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [ ] Gallery example in `./doc/examples` (new features only)
- [ ] Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- [ ] Unit tests
- [ ] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

<!-- For detailed information on these and other aspects see -->
<!-- the scikit-image contribution guidelines. -->
<!-- https://scikit-image.org/docs/dev/contribute.html -->

## For reviewers

<!-- Don't remove the checklist below. -->
- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

pep8speaks commented:
Hello @m-albert! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`doc/examples/color_exposure/plot_adapt_hist_eq.py`](https://github.com/scikit-image/scikit-image/blob/888959cc51ef120bc3f77d6807d47b5293960997/doc/examples/color_exposure/plot_adapt_hist_eq.py):

> [Line 6:80](https://github.com/scikit-image/scikit-image/blob/888959cc51ef120bc3f77d6807d47b5293960997/doc/examples/color_exposure/plot_adapt_hist_eq.py#L6): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)
> [Line 7:80](https://github.com/scikit-image/scikit-image/blob/888959cc51ef120bc3f77d6807d47b5293960997/doc/examples/color_exposure/plot_adapt_hist_eq.py#L7): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 8:80](https://github.com/scikit-image/scikit-image/blob/888959cc51ef120bc3f77d6807d47b5293960997/doc/examples/color_exposure/plot_adapt_hist_eq.py#L8): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 87:15](https://github.com/scikit-image/scikit-image/blob/888959cc51ef120bc3f77d6807d47b5293960997/doc/examples/color_exposure/plot_adapt_hist_eq.py#L87): [E126](https://duckduckgo.com/?q=pep8%20E126) continuation line over-indented for hanging indent

emmanuelle commented:
Hi @m-albert thank you for the PR! The Travis CI is currently failing for the builds using minimal requirements, I believe that it's related to matplotlib. Could you please take a look? Thanks :-).

m-albert commented:
Hey @emmanuelle, thanks for making me aware of it :) Should be fixed.

alexdesiqueira commented:
Hey @m-albert, thank you very much for this — it looks great!
I have a suggestion, though. Could we use an _actual_ image, or a more well-behaved set, as an example? For me it's kinda hard to see the AHE effects on this set.
Thanks again!

m-albert commented:
Hey @alexdesiqueira, thanks for your suggestion!

I do agree that the image is not super clear in the sense that the viewer does not immediately recognize much in it. However it could exemplify e.g. a dense measurement of something in 3d. And I thought including the degradation along x would help the illustration...

Would you have a suggestion for a better 3D image? Or could you point me to available example volumes within scikit-image?

Alternatively, I can think of these ways of making the illustration clearer:
1) including non-adaptive histogram equalization for comparison and/or
2) choosing a different random seed.

That could look something like this:

![image](https://user-images.githubusercontent.com/12528388/82154814-cb224000-9870-11ea-8e5d-0a9ad670f0de.png)

alexdesiqueira commented:
@m-albert thank you, that looks great already! More context surely made it better.
I was thinking on an image in the lines of [cells.tif](https://github.com/scikit-image/skimage-tutorials/blob/master/images/cells.tif), which we use on skimage 3D tutorials. It's a large image, though; maybe a slice of it would suffice... what do you think, @emmanuelle?
Thanks again!

jni commented:
@m-albert @alexdesiqueira I actually disagree with Alex here, I think this example shows very nicely how AHE is able to give you a similar image for both degraded and non-degraded images, starting with a *very* degraded image. If we add cells.tif, I would suggest using it *in addition to* the synthetic example.

On the other hand, we could also use cells.tif together with some synthetic degradation...

alexdesiqueira commented:
@m-albert never mind me then; @jni's idea is better!

m-albert commented:
@alexdesiqueira I wasn't aware of the cell dataset, I think that's a great suggestion!

> On the other hand, we could also use cells.tif together with some synthetic degradation...


@jni Great idea, here's a combination of the cells with synthetic degradation. What do you think guys?

![image](https://user-images.githubusercontent.com/12528388/82248526-95508a80-9948-11ea-8852-8eaa6c79a529.png)

alexdesiqueira commented:
@m-albert this looks amazing! Thank you!

jni commented:
> What do you think guys?

:heart_eyes: :heart_eyes: :heart_eyes: :heart_eyes: :heart_eyes:

m-albert commented:
I'm just realizing that if skimage-tutorials can be used for stably hosting images I could have uploaded a zebrafish embryo like one of these ones: https://twitter.com/i/status/1259893967681916930. Maybe for the next example :)

m-albert commented:
Hmm some of the checks fail but I don't understand why. The output appears unrelated to my added file. What can I do? Thanks for any help :)

```
=========================== short test summary info ===========================
FAILED python37-x64/lib/site-packages/skimage/color/tests/test_colorconv.py::TestColorconv::test_lab_full_gamut
FAILED python37-x64/lib/site-packages/skimage/color/tests/test_colorconv.py::TestColorconv::test_rgb2gray_alpha
FAILED python37-x64/lib/site-packages/skimage/color/tests/test_colorconv.py::test_nD_gray_conversion[shape0-rgb2gray]
FAILED python37-x64/lib/site-packages/skimage/color/tests/test_colorconv.py::test_nD_gray_conversion[shape2-rgb2gray]
FAILED python37-x64/lib/site-packages/skimage/color/tests/test_colorconv.py::test_nD_gray_conversion[shape3-rgb2gray]
FAILED python37-x64/lib/site-packages/skimage/color/tests/test_colorconv.py::test_nD_gray_conversion[shape4-rgb2gray]
FAILED python37-x64/lib/site-packages/skimage/feature/tests/test_corner.py::test_shape_index
FAILED python37-x64/lib/site-packages/skimage/filters/tests/test_gaussian.py::test_multichannel
FAILED python37-x64/lib/site-packages/skimage/filters/tests/test_ridges.py::test_border_warning[sato]
FAILED python37-x64/lib/site-packages/skimage/filters/tests/test_ridges.py::test_border_warning[hessian]
FAILED python37-x64/lib/site-packages/skimage/io/tests/test_pil.py::test_all_color
FAILED python37-x64/lib/site-packages/skimage/io/tests/test_pil.py::test_all_mono
FAILED python37-x64/lib/site-packages/skimage/restoration/tests/test_denoise.py::test_denoise_bilateral_nan
FAILED python37-x64/lib/site-packages/skimage/segmentation/tests/test_random_walker.py::test_2d_bf
FAILED python37-x64/lib/site-packages/skimage/segmentation/tests/test_random_walker.py::test_2d_cg
FAILED python37-x64/lib/site-packages/skimage/segmentation/tests/test_random_walker.py::test_2d_cg_mg
FAILED python37-x64/lib/site-packages/skimage/segmentation/tests/test_random_walker.py::test_2d_cg_j
FAILED python37-x64/lib/site-packages/skimage/segmentation/tests/test_random_walker.py::test_types
FAILED python37-x64/lib/site-packages/skimage/segmentation/tests/test_random_walker.py::test_reorder_labels
FAILED python37-x64/lib/site-packages/skimage/segmentation/tests/test_random_walker.py::test_2d_inactive
FAILED python37-x64/lib/site-packages/skimage/segmentation/tests/test_random_walker.py::test_3d
FAILED python37-x64/lib/site-packages/skimage/segmentation/tests/test_random_walker.py::test_3d_inactive
FAILED python37-x64/lib/site-packages/skimage/segmentation/tests/test_random_walker.py::test_multispectral_2d
FAILED python37-x64/lib/site-packages/skimage/segmentation/tests/test_random_walker.py::test_multispectral_3d
FAILED python37-x64/lib/site-packages/skimage/segmentation/tests/test_random_walker.py::test_spacing_0
FAILED python37-x64/lib/site-packages/skimage/segmentation/tests/test_random_walker.py::test_spacing_1
FAILED python37-x64/lib/site-packages/skimage/segmentation/tests/test_random_walker.py::test_trivial_cases
FAILED python37-x64/lib/site-packages/skimage/segmentation/tests/test_random_walker.py::test_length2_spacing
FAILED python37-x64/lib/site-packages/skimage/segmentation/tests/test_random_walker.py::test_isolated_seeds
FAILED python37-x64/lib/site-packages/skimage/segmentation/tests/test_random_walker.py::test_prob_tol
FAILED python37-x64/lib/site-packages/skimage/transform/tests/test_radon_transform.py::test_radon_circle
FAILED python37-x64/lib/site-packages/skimage/transform/tests/test_warps.py::test_swirl
FAILED python37-x64/lib/site-packages/skimage/transform/tests/test_warps.py::test_downsize_anti_aliasing_invalid_stddev
FAILED python37-x64/lib/site-packages/skimage/transform/tests/test_warps.py::test_boll_array_warnings
== 34 failed, 5828 passed, 163 skipped, 18289 warnings in 130.10s (0:02:10) ===
Command exited with code 1
```

rfezzani commented:
These CI failures are fixed by #4731 and #4735, you need to rebase if you want an all green CI ;)

m-albert commented:
Oh I see, thanks a lot @rfezzani :)

rfezzani commented on doc/examples/color_exposure/plot_adapt_hist_eq_3d.py:
```diff
@@ -0,0 +1,246 @@
+
+# highlight AHE
+rect_ax = fig.add_axes([0, 0, 1, 1], facecolor='none')
+rect_ax.axis(False)
```
```suggestion
rect_ax.set_axis_off()
```
This line is causing CI failures ;)

m-albert replied:
Thanks!

rfezzani commented on doc/examples/color_exposure/plot_adapt_hist_eq_3d.py:
```diff
@@ -0,0 +1,246 @@
+
+import numpy as np
+from skimage import exposure
+from skimage import io
```
Please use `imageio` instead.

rfezzani commented on doc/examples/color_exposure/plot_adapt_hist_eq_3d.py:
```diff
@@ -0,0 +1,246 @@
+# Prepare data and apply hist. eq.
+#############
+
+im_orig = io.imread('https://github.com/scikit-image/'
+                    'skimage-tutorials/blob/master/images/cells.tif?raw=True',
+                    plugin='tifffile')
```
```suggestion
im_orig = imageio.volread('https://github.com/scikit-image/'
                    'skimage-tutorials/blob/master/images/cells.tif?raw=True')
```

m-albert commented:
PR should be ready from my side. Thanks a lot @rfezzani for reviewing!

emmanuelle commented on doc/examples/color_exposure/plot_adapt_hist_eq_3d.py:
```diff
@@ -0,0 +1,247 @@
+# Prepare data and apply hist. eq.
+#############
+
+im_orig = imageio.volread('https://github.com/scikit-image/'
+                          'skimage-tutorials/blob/master/'
+                          'images/cells.tif?raw=True')
```
```suggestion
from skimage.data import image_fetcher
path = image_fetcher.fetch('data/cells.tif')
img_orig = io.imread(path) 
```

emmanuelle replied:
this way we can use our new fetching mechanism with pooch, and if the data have already been downloaded once, they'll be cached for the next time they are opened.

emmanuelle commented:
I just left a comment about I/O, if you agree with it @m-albert please accept it and then I'll merge. Thanks for a great example!

emmanuelle commented:
You need to import `io` in the gallery example (see CI failure).

m-albert commented:
Thanks for your review @emmanuelle! I incorporated your suggestion and fixed some `matplotlib` issues that started to show up with its 3.3 release that came out in the meantime.

And I think the tests are as green as I can get them after rebasing :)

soupault commented on doc/examples/color_exposure/plot_adapt_hist_eq_3d.py:
```diff
@@ -0,0 +1,250 @@
+
+# add boxes illustrating the kernels
+ns = np.array(im_orig.shape) // kernel_size - 1
+for iax, ind, rect_shape in zip([1] + [2] * 4,
```
`iax, ind` could be renamed to improve readibility.

m-albert replied:
Renamed to `axis_ind, vertex_ind, box_shape`, hope that's clearer.

soupault commented on doc/examples/color_exposure/plot_adapt_hist_eq_3d.py:
```diff
@@ -0,0 +1,250 @@
+    ax.set_ylim3d(XYZlim)
+    ax.set_zlim3d(XYZlim * 0.55)
+
+    try:
```
Is this neccessary?

m-albert replied:
I think so, see this thread (commented in the code) https://github.com/matplotlib/matplotlib/issues/17172#issuecomment-617546105. Definitely `napari` will be better for 3D plotting!

soupault commented on doc/examples/color_exposure/plot_adapt_hist_eq_3d.py:
```diff
@@ -0,0 +1,250 @@
+    ax.elev = 30
+
+
+plt.subplots_adjust(left=0.01,
```
Can be shortened to 1-2 lines.

soupault commented on doc/examples/color_exposure/plot_adapt_hist_eq_3d.py:
```diff
@@ -0,0 +1,250 @@
+im_orig = io.volread(path)
+im_orig = im_orig.transpose()
+
+# apply some rescaling
```
`some` is somewhat not informative. I would change to `Apply rescaling`, `Rescale image intensity from 0 to 1`, etc.

m-albert replied:
Good point.

m-albert commented on doc/examples/color_exposure/plot_adapt_hist_eq_3d.py:
```diff
@@ -0,0 +1,250 @@
+
+#############
+# Prepare data and apply hist. eq.
+#############
```
True, the hashtags can be misunderstood for sphinx formatting

m-albert commented on doc/examples/color_exposure/plot_adapt_hist_eq_3d.py:
```diff
@@ -0,0 +1,250 @@
+# define functions to help plot the data
+#############
+
+def get_rgba(scalars, cmap, vmin=0., vmax=1., alpha=0.2):
```
Definitely.

m-albert commented:
Hey @soupault, thanks a lot for your review and feedback! I incorporated all of your suggestions for the code.

> For the figure, I think it would be clearer if the captions from the left edge are put above the subfigures. Additionally, the captions for the bottom right subfigures would be helpful to better differentate the first row from the second. Summarized:

Thanks for the detailed cosmetic suggestions :) I played around a little with caption rearrangements and think that
- for now I prefer the tabular format (at least I couldn't come up with individual captions that I'd feel would increase readability)
- as you suggest, stating more clearly what the input is and how it becomes modified definitely improves the figure (that might also differentiate a bit more the first and second rows).

Let me know what you think!

In unrelated matters, something must have changed in `matplotlib` 3.3. as I'm getting changes in subfigure arrangement and some 3d properties.. hope that'll stay relatively stable in the near future.

![image](https://user-images.githubusercontent.com/12528388/88199494-390a2f80-cc45-11ea-83d7-0a377c0e761b.png)

rfezzani commented:
I agree with @soupault, what a piece of work @m-albert !! Thank you very much for your amazing contribution :clap:

m-albert commented:
Thanks for all the feedback, encouragement and great vibes guys :)

jni commented:
I thought this had merged ages ago! Oops! 😅 Better late than never and thank you yet again @m-albert! 🎉🎉
